src/internal/rest.rs

```diff
@@ -36,69 +44,273 @@ pub struct ServerError {
+/// We sign over: `/api/1/users?id=abcABC012_.%24%23%7C%40%2F%3A%3B%3D%2B%27-`
+///
+/// Anyone wanting to verify this signature will need to be able to match this exact encoding.
+#[derive(Debug, Clone)]
+pub struct SignatureUrlString(String);
+
+impl SignatureUrlString {
+    pub fn new(encoded_full_url: &str) -> Result<SignatureUrlString, reqwest::UrlError> {
+        let parsed_url = Url::parse(encoded_full_url)?;
+        let query_str_format = |q: &str| format!("?{}", q);
+
+        Ok(SignatureUrlString(format!(
+            "{}{}",
+            parsed_url.path(),
+            parsed_url.query().map_or("".into(), query_str_format)
+        )))
+    }
+
+    /// String to sign over
+    fn signature_string(&self) -> &str {
+        &self.0
+    }
+}
+
+/// Representation of X-IronCore-User-Context header
+#[derive(Clone, Debug)]
+pub struct HeaderIronCoreUserContext {
+    timestamp: DateTime<Utc>,
+    segment_id: usize,
+    user_id: UserId,
+    public_signing_key: [u8; 32],
+}
+
+impl HeaderIronCoreUserContext {
+    /// Payload of the header
+    fn payload(&self) -> String {
+        format!(
+            "{},{},{},{}",
+            self.timestamp.timestamp_millis(),
+            self.segment_id,
+            self.user_id.id(),
+            base64::encode(&self.public_signing_key)
+        )
+    }
+
+    /// Signature over the header's payload
+    fn signature(&self, signing_keys: &DeviceSigningKeyPair) -> [u8; 64] {
+        signing_keys.sign(&self.payload().into_bytes())
+    }
+
+    /// To a reqwest-compatible header
+    fn to_header(&self, error_code: RequestErrorCode) -> Result<HeaderMap, IronOxideErr> {
+        let mut headers: HeaderMap = Default::default();
+        self.payload()
+            .parse()
+            .map_err(|_| IronOxideErr::RequestError {
+                message: format!(
+                    "Failed to encode '{}' into a X-IronCore-User-Context header",
+                    &self.payload()
+                ),
+                code: error_code,
+                http_status: None,
+            })
+            .map(|url| {
+                headers.append("X-IronCore-User-Context", url);
+                headers
+            })
+    }
+}
+
+/// Representation of X-IronCore-Request-Sig header
+#[derive(Clone, Debug)]
+pub struct HeaderIronCoreRequestSig<'a> {
+    ironcore_user_context: HeaderIronCoreUserContext,
+    method: Method,
+    url: SignatureUrlString,
+    body: Option<&'a [u8]>, //serialization of this body has to be identical to that in IronCoreRequest
+    signing_keys: &'a DeviceSigningKeyPair,
+}
+
+impl<'a> HeaderIronCoreRequestSig<'a> {
+    /// Payload of the header
+    fn payload(&self) -> Vec<u8> {
+        let HeaderIronCoreRequestSig {
+            body,
+            ironcore_user_context,
+            method,
+            url,
+            ..
+        } = self;
+
+        // use closure here to delay computation until we know if we need to append the body or not
```

Comment:
I experimented a little:
```rust
        let bytes_no_body = format!(
            "{}{}{}",
            &ironcore_user_context.payload(),
            &method,
            url.signature_string(),
        )
        .into_bytes();

        body.map_or_else(
            || bytes_no_body.clone(),
            |body_bytes| [&bytes_no_body, body_bytes].concat(),
        )
```
This has an extra `clone()` needed because the closure captures the value. I could also duplicate the format expression, but that doesn't seem better. I did clean it up a little, but I think the closure is still a good solution.

Reply:
If you use a `match` you can use it in both branches without the clone I think. `match body {...}`